src/compiler/checker.ts

```diff
@@ -3236,6 +3237,51 @@ namespace ts {
             return undefined;
         }
 
+        /**
+         * Gets combined flags of a `symbol` and all alias targets it resolves to. `resolveAlias`
+         * is typically recursive over chains of aliases, but stops mid-chain if an alias is merged
+         * with another exported symbol, e.g.
+         * ```ts
+         * // a.ts
+         * export const a = 0;
+         * // b.ts
+         * export { a } from "./a";
+         * export type a = number;
+         * // c.ts
+         * import { a } from "./b";
+         * ```
+         * Calling `resolveAlias` on the `a` in c.ts would stop at the merged symbol exported
+         * from b.ts, even though there is still more alias to resolve. Consequently, if we were
+         * trying to determine if the `a` in c.ts has a value meaning, looking at the flags on
+         * the local symbol and on the symbol returned by `resolveAlias` is not enough.
+         * @returns undefined if `symbol` is an alias that ultimately resolves to `unknown`;
+         * combined flags of all alias targets otherwise.
+         */
+        function getAllSymbolFlags(symbol: Symbol): SymbolFlags | undefined {
+            let flags = symbol.flags;
+            let seenSymbols;
+            while (symbol.flags & SymbolFlags.Alias) {
+                const target = resolveAlias(symbol);
+                if (target === unknownSymbol) {
+                    return undefined;
+                }
+
+                // Optimization - avoid creating `seenSymbols` Set in most cases
+                if (target === symbol) {
+                    break;
+                }
+                else if (target.flags & SymbolFlags.Alias && !seenSymbols) {
+                    seenSymbols = new Set([symbol, target]);

```

Comment:
This is the only place where `seenSymbols` is populated with values; does this need to be something like

```ts
  else if (target.flags & SymbolFlags.Alias) {
    if (!seenSymbols) {
      seenSymbols = new Set([symbol, target]);
    } else {
      seenSymbols.add(target);
    }
  }
```